ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -996,12 +1014,76 @@ let rank_hosts_by_best_vgpu ~__context vgpu visible_hosts =
                0L
         )
         hosts
-      |> List.map (fun g -> List.map (fun (h, _) -> h) g)
+      |> List.map (fun g -> List.map fst g)
+
+(* Group all hosts to 2 parts:
+   1. A list of affinity host (only one host).
+   2. A list of lists, each list contains hosts with the same number of
+      running VM in that anti-affinity group.
+      These lists are sorted by VM's count.
+   Combine these lists into one list. The list is like below:
+   [ [host1] (affinity host)
+   , [host2, host3] (no VM running)
+   , [host4, host5] (one VM running)
+   , [host6, host7] (more VMs running)
+   , ...
+   ]
+*)
+let rank_hosts_by_placement ~__context ~vm ~group =
+  let affinity_host =
+    match Db.VM.get_affinity ~__context ~self:vm with
+    | ref when Db.is_valid_ref __context ref ->
+        [ref]
+    | _ ->
+        []
+  in
+  let sorted_hosts =
+    let host_without_affinity_host =
+      Db.Host.get_all ~__context
+      |> List.filter (fun host -> not (List.mem host affinity_host))
+    in
+    let host_tbl =
+      host_without_affinity_host
+      |> List.map (fun host -> (host, 0))
+      |> List.to_seq
+      |> Hashtbl.of_seq
+    in
+    Db.VM_group.get_VMs ~__context ~self:group
+    |> List.filter_map (fun vm ->
+           let vm_rec = Db.VM.get_record ~__context ~self:vm in
+           match
+             ( vm_rec.API.vM_power_state
+             , vm_rec.API.vM_resident_on
+             , vm_rec.API.vM_scheduled_to_be_resident_on
+             )
+           with
+           | `Running, h, _ when h <> Ref.null ->
```

Comment:
When a VM start migrating, it's `scheduled_to_be_resident_on` will be set, while its `resident_on` is not cleared?
Then it should be:
`| `Running, h, sh when h <> Ref.null && sh = Ref.null`

Reply:
Like this?
`           match
             ( vm_rec.API.vM_power_state
             , vm_rec.API.vM_scheduled_to_be_resident_on
             , vm_rec.API.vM_resident_on
             )
           with
           | _, sh, _ when sh <> Ref.null ->
               Some sh
           | `Running, _, h when h <> Ref.null ->
               Some h
           | _ ->
               None
`